src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,276 @@
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Runtime.InteropServices;
+using System.IO;
+using System.Threading.Tasks;
+using System.Collections.Concurrent;
+
+namespace System.Device.Gpio.Drivers
+{
+    public class LibGpiodDriver : UnixDriver
+    {
+        private SafeChipHandle _chip;
+
+        private Dictionary<int, SafeLineHandle> _pinNumberToSafeLineHandle;
+
+        private ConcurrentDictionary<int, LibGpiodDriverEventHandler> _pinNumberToEventHandler = new ConcurrentDictionary<int, LibGpiodDriverEventHandler>();
+
+        protected internal override int PinCount => Interop.GetNumberOfLines(_chip);
+
+        public LibGpiodDriver()
+        {
+            OpenChip();
+            _pinNumberToSafeLineHandle = new Dictionary<int, SafeLineHandle>(PinCount);
+        }
+
+        private void OpenChip()
+        {
+            SafeChipIteratorHandle iterator = Interop.GetChipIterator();
+            if (iterator == null)
+            {
+                throw new IOException($"No any chip available, error code: {Marshal.GetLastWin32Error()}");
+            }
+            _chip = Interop.GetNextChipFromChipIterator(iterator);
+            if (_chip == null)
+            {
+                throw new IOException($"No chip found, error code: {Marshal.GetLastWin32Error()}");
+            }
+            // Freeing other chips opened
+            Interop.FreeChipIteratorNoCloseCurrentChip(iterator);
+        }
+
+        private SafeLineHandle RequestEvent(int pinNumber, PinEventTypes eventType)
+        {
+            int eventSuccess = -1;
+            SafeLineHandle pin = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pin != null)
+            {
+                if (eventType == PinEventTypes.Rising || eventType == PinEventTypes.Falling)
+                {
+                    eventSuccess = Interop.RequestBothEdgeEventForLine(pin, $"Listen {pinNumber} for falling edge event");
+                }
+                else
+                {
+                    throw new ArgumentException("Invalid or not supported event type requested");
+                }
+            }
+            if (eventSuccess < 0)
+            {
+                throw new IOException($"Error while requesting event listener for pin {pinNumber}, error code: {Marshal.GetLastWin32Error()}");
+            }
+            return pin;
+        }
+
+        protected internal override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventType, PinChangeEventHandler callback)
+        {
+            if (!_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
+            {
+                if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+                    Interop.ReleaseGpiodLine(pin);
+                eventHandler = new LibGpiodDriverEventHandler(pinNumber, new CancellationTokenSource());
```

Comment:
Do you want to keep the CancellationTokenSource around? I would do that and that way you can always just send a cancel request. Unless you arleady do that in your event handler.

Reply:
Not sure what you mean by to keep it around, but JFYI it is used here: https://github.com/dotnet/iot/pull/133/files#diff-47ebc9a36c6ca0979542276c920324f9R177